### PR TITLE
Fix M1 tail call issue when building

### LINKED ISSUE #44107: Building master branch crashes on Apple M1
is building current Julia master crashing for anyone else? (i.e. commit aae68a5e030a04f3d4bfc31f4b30c002ddbc52ca, on : macOS 12.2 (21D49) on Apple M1 Max, Xcode 13.2.1).

It builds fine for me on Ubuntu x86.

```
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 36/36
Executing precompile statements... 1962/2006
Precompilation complete. Summary:
Total ───────  71.210485 seconds
Generation ──  52.778831 seconds 74.1167%
Execution ───  18.431653 seconds 25.8833%
LLVM ERROR: failed to perform tail call elimination on a call site marked musttail

signal (6): Abort trap: 6
in expression starting at none:0
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 158165722 (Pool: 158088219; Big: 77503); GC: 157
/bin/sh: line 1: 62460 Abort trap: 6           JULIA_BINDIR=/Users/mhorn/Projekte/Julia/julia.master/usr/bin WINEPATH="/Users/mhorn/Projekte/Julia/julia.master/usr/bin;$WINEPATH" /Users/mhorn/Projekte/Julia/julia.master/usr/bin/julia -O3 -C "apple-a12" --output-o /Users/mhorn/Projekte/Julia/julia.master/usr/lib/julia/sys-o.a.tmp --startup-file=no --warn-overwrite=yes --sysimage /Users/mhorn/Projekte/Julia/julia.master/usr/lib/julia/sys.ji /Users/mhorn/Projekte/Julia/julia.master/contrib/generate_precompile.jl 1
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/mhorn/Projekte/Julia/julia.master/usr/lib/julia/sys-o.a] Error 1
make: *** [julia-sysimg-release] Error 2
```

### PR DESCRIPTION
**Pull request by gbaraldi**
This implements @staticfloat solution for the LLVM error that appears on m1 builds #44107.
Current master is very brittle when building the m1 but this bug happens every time without the PR.
I tried debugging it quite a lot and LLVM DEBUG with arm lower call debug on doesn't show why it doesn't tail call so I have no idea why it fails, the bug was added by the purity PR as mentioned in the issue. I don't know if it has to do with the anonymous union but apple gives a warning for that

### CODE REVIEW

**Review by navidcy — APPROVED** [review]
This works on my MacBook Pro M1 Max.

**Comment by navidcy** [comment]
```Julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.9.0-DEV.35 (2022-02-20)
 _/ |\__'_|_|_|\__'_|  |  fix-m1-build/e102b5caaf (fork: 63 commits, 5 days)
|__/                   |
```

File: src/ccall.cpp [hunk 1]
```diff
@@ -233,10 +233,11 @@ static GlobalVariable *emit_plt_thunk(
     else {
         // musttail support is very bad on ARM, PPC, PPC64 (as of LLVM 3.9)
         // Known failures includes vararg (not needed here) and sret.
```
Comment (vtjnash) [inline src/ccall.cpp:235]: possible improvement?
```suggestion
        // Known failures includes vararg (not needed here) and sret.
        ret->setTailCall();
```

Is this only M1, or all AArch64?

Comment (gbaraldi) [reply src/ccall.cpp:235]: This is m1 only.  I wonder if it's this https://github.com/llvm/llvm-project/blob/d1cd64ffdd832220dbe1829c2f09b880be67be31/llvm/test/CodeGen/AArch64/GlobalISel/call-translator-musttail.ll, this is the only aarch64-apple exclusive thing related to musttail.

That might be a more robust solution for sure. Does doing the ccall as a tail call have some performance benefit?

**Comment by gbaraldi** [comment]
I found what causes musttail to fail, it's that the stack arguments of the call don't fit into the save area for the callee. https://github.com/gbaraldi/llvm-project/blob/5be8ea4bd7b0d66214c22033a7fd298544e65027/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L5702-L5707. Not sure if it's something to be worried or not however.

**Comment by KristofferC** [comment]
I'll merge this now since it fixes the build and we want this for 1.8-beta to be available on M1. https://github.com/JuliaLang/julia/pull/44279#discussion_r812109595 can be discussed further and possible improvements can be done in future PRs.

**Comment by ngam** [comment]
@KristofferC, I recommend you keep track of @gbaraldi investigation somewhere to ensure the work doesn't get lost. Yes, merging it is fine and preferable --- thanks for that! But maybe open a new issue to keep track of all the development. (Maybe @gbaraldi can do that?)

Thanks a lot for the great work!!

**Comment by gbaraldi** [comment]
I will open an issue to keep this is mind, this is is a consequence of the purity PR, there are some other issues related to it too so I might open a more general issue.

**Pull request merged**